parlai/tasks/squad/agents.py

```diff
@@ -15,62 +15,116 @@
+        self.random = self.datatype == 'train'
+
+        # for ordered data in batch mode (especially, for validation and
+        # testing), each teacher in the batch gets a start index and a step
+        # size so they all process disparate sets of the data
+        self.step_size = opt.get('batchsize', 1)
+        self.data_offset = opt.get('batchindex', 0)
+        self.reset()
 
     def __len__(self):
-        return self.len
+        return len(self.examples)
+
+    def reset(self):
+        """Reset the dialog so that it is at the start of the epoch,
+        and all metrics are reset.
+        """
+        self.metrics.clear()
+        self.lastY = None
+        self.episode_idx = self.data_offset - self.step_size
+        self.episode_done = True
+        self.epochDone = False
+        if not self.random and self.data_offset >= self.data.num_episodes():
+            # could have bigger batchsize then episodes... so nothing to do
+            self.epochDone = True
+
+    def observe(self, observation):
+        """Process observation for metrics."""
+        if self.lastY is not None:
+            self.metrics.update(observation, self.lastY)
+            self.lastY = None
+        return observation
 
     # return state/action dict based upon passed state
     def act(self):
-        if self.datatype == 'train':
-            self.episode_idx = random.randrange(len(self.examples))
+        """Send new dialog message."""
+        if self.epochDone:
+            return {'id': 'squad', 'episode_done': True}
+
+        num_eps = len(self.examples)
+
+        if self.random:
+            # select random episode
+            self.episode_idx = random.randrange(num_eps)
         else:
-            self.episode_idx = (self.episode_idx + 1) % len(self.examples)
+            # select next episode
+            self.episode_idx = (self.episode_idx + self.step_size) % num_eps
+
         article_idx, paragraph_idx, qa_idx = self.examples[self.episode_idx]
         article = self.squad[article_idx]
         paragraph = article['paragraphs'][paragraph_idx]
         qa = paragraph['qas'][qa_idx]
         question = qa['question']
-        answers = [a['text'] for a in qa['answers']]
+        answers = []
+        answer_starts = []
+        for a in qa['answers']:
+            answers.append(a['text'])
+            answer_starts.append(a['answer_start'])
         context = paragraph['context']
 
-        if (self.episode_idx == (len(self.examples) - 1) and
-            self.datatype != 'train'):
+        if not self.random and self.episode_idx + self.step_size >= num_eps:
+            # this is used for ordered data to check whether there's more data
             self.epochDone = True
 
-        return {
+        action = {
+            'id': 'squad',
             'text': context + '\n' + question,
-            'labels': answers,
-            'episode_done': True
+            'episode_done': True,
+            'answer_starts': answer_starts
```

Comment:
hmm..slightly weird name but i guess that's what they called it... why plural though?

Reply:
one start for each label :) (there are several labels for every question)